<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Fi-NeMo Report</title>
    <style>
        /* Adapted from https://github.com/oxalorg/sakura/ */
        html {
            font-size: 62.5%;
            font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif;
        }
        
        body {
            font-size: 1.8rem;
            line-height: 1.618;
            max-width: 60em;
            margin: auto;
            color: #222222;
            background-color: #ffffff;
            padding: 13px;
        }
        
        @media (max-width: 684px) {
        body {
            font-size: 1.53rem;
        }
        }
        @media (max-width: 382px) {
        body {
            font-size: 1.35rem;
        }
        }
        h1, h2, h3, h4, h5, h6 {
            line-height: 1.1;
            font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif;
            font-weight: 700;
            margin-top: 3rem;
            margin-bottom: 1.5rem;
            overflow-wrap: break-word;
            word-wrap: break-word;
            -ms-word-break: break-all;
            word-break: break-word;
        }
        
        h1 {
            font-size: 2.35em;
        }
        
        h2 {
            font-size: 1.75em;
        }
        
        h3 {
            font-size: 1.25em;
        }
        
        p {
            margin-top: 0px;
            margin-bottom: 2.5rem;
        }
        
        small, sub, sup {
            font-size: 75%;
        }
        
        hr {
            border-color: #007559;
        }
        
        a {
            text-decoration: none;
            color: #007559;
        }
        a:visited {
            color: #004232;
        }
        a:hover {
            color: #006994;
            border-bottom: 2px solid #222222;
        }
        
        ul {
            padding-left: 1.4em;
            margin-top: 0px;
            margin-bottom: 2.5rem;
        }
        
        li {
            margin-bottom: 0.4em;
        }
        
        blockquote {
            margin-left: 0px;
            margin-right: 0px;
            padding-left: 1em;
            padding-top: 0.8em;
            padding-bottom: 0.8em;
            padding-right: 0.8em;
            border-left: 5px solid #007559;
            margin-bottom: 2.5rem;
            background-color: #f7f7f7;
        }
        
        blockquote p {
            margin-bottom: 0;
        }
        
        img, video {
            height: auto;
            max-width: 100%;
            margin-top: 0px;
            margin-bottom: 2.5rem;
        }
        
        /* Pre and Code */
        pre {
            background-color: #f7f7f7;
            display: block;
            padding: 1em;
            overflow-x: auto;
            margin-top: 0px;
            margin-bottom: 2.5rem;
            font-size: 0.9em;
        }
        
        code, kbd, samp {
            font-size: 0.9em;
            padding: 0 0.5em;
            background-color: #f7f7f7;
            white-space: pre-wrap;
            }
        
        pre > code {
            padding: 0;
            background-color: transparent;
            white-space: pre;
            font-size: 1em;
        }
        
        /* Tables */
        table {
            width: 100%;
            border-collapse: collapse;
            margin-bottom: 2rem;
            font-variant-numeric: tabular-nums;
            text-align: justify;
        }

        td, th {
            padding: 0.5em;
            border-bottom: 1px solid #f1f1f1;
        }

        .num_col {
            text-align: right;
        }

        td.num_col {
            padding-left: 2em;
        }

        th.cwm, th.distplot {
            padding-left: 2em;
        }

        .cwm img {
            max-width: unset;
            margin-bottom: 0
        }

        .distplot img {
            max-width: unset;
            margin-bottom: 0
        }
        
    </style>
</head>

<body>
<h1>Fi-NeMo hit calling report</h1>

{% if not use_seqlets %}
<blockquote>
    Seqlet comparisons are not shown because a TF-MoDISco H5 file with seqlet data is not provided.
</blockquote>
{% elif not compute_recall %}
<blockquote>
    Seqlet recall and other statistics directly comparing hits and seqlets are not computed because the <code>-n/--no-recall</code> argument is set.
</blockquote>
{% endif %}

{% if use_seqlets %}
<h2>TF-MoDISco seqlet comparisons</h2>
<p>
    The following figures and statistics compare the called hits with the seqlets used by TF-MoDISco to construct each motif. 
</p>

<h3>Hit vs. seqlet counts</h3>
<p>
    This figure shows the number of hits called vs. the number of TF-MoDISco seqlets identified for each motif. 
    The dashed line is the identity line. 
    When comparing a shared set of regions, the hit counts should be mostly greater than the corresponding seqlet counts, since TF-MoDISco stringently filters seqlets and usually uses a smaller input window. 
</p>
<img src="hit_vs_seqlet_counts.svg" width="780">
{% endif %}

<h3>Hit and seqlet motif comparisons</h3>
<p>
    For each motif, this table examines the consistency between hits and TF-MoDISco seqlets. 
</p>

<p>
    The following statistics report the number of hits, seqlets, and their relationships: 
</p>
<ul>
    <li><strong>Hits</strong>: The number of hits called by Fi-NeMo</li>
    {% if compute_recall %}
    <li><strong>Restricted Hits</strong>: The number of Fi-NeMo hits within the TF-MoDISco input regions</li>
    {% endif %}
    {% if use_seqlets %}
    <li><strong>Seqlets</strong>: The number of unique TF-MoDISco seqlets</li>
    {% endif %}
    {% if compute_recall %}
    <li><strong>Hit/Seqlet Overlaps</strong>: The number of hits that coincide with TF-MoDISco seqlets</li>
    <li><strong>Missed Seqlets</strong>: The number of TF-MoDISco seqlets not called as hits</li>
    <li><strong>Additional Restricted Hits</strong>: The number of hits within the TF-MoDISco input regions that are not identified as seqlets by TF-MoDISco</li>
    <li><strong>Seqlet Recall</strong>: The fraction of seqlets that are called as hits</li>
    {% endif %}
    <li><strong>Hit-Seqlet CWM Similarity</strong>: The cosine similarity between the hit CWM and the TF-MoDISco CWM</li>
</ul>
<p>
    Note that the seqlet counts here may be lower than those shown in the tfmodisco-lite report due to double-counting in overlapping regions.
    The seqlet counts shown here are unique while the counts in the tfmodisco-lite report are not de-duplicated.
</p>
{% if compute_recall %}
<p>
    Note that palindromic motifs may have lower recall due to disagreements on orientation.
    If seqlet recall is near zero for all motifs, the <code>-W/--modisco-region-width</code> argument is likely incorrect.
    This value is required to infer genomic coordinates of seqlets from the tfmodisco-lite output H5.
</p>
{% endif %}
<p>
    Motif CWMs (contribution weight matrices) are average contribution scores over a set of regions. The CWMs plotted here are:
</p>
<ul>
    <li><strong>Hit CWM (FC)</strong>: The forward-strand CWM of all hits</li>
    <li><strong>Hit CWM (RC)</strong>: The reverse-strand CWM of all hits</li>
    <li><strong>TF-MoDISco CWM (FC/RC)</strong>: The CWM of all TF-MoDISco seqlets</li>
    {% if compute_recall %}
    <li><strong>Missed-Seqlet-Only CWM</strong>: The CWM of all TF-MoDISco seqlets that were not called as hits</li>
    <li><strong>Additional-Restricted-Hit CWM</strong>: The CWM of all hits within the TF-MoDISco input regions that were not identified as seqlets by TF-MoDISco</li>
    {% endif %}
</ul>
<p>
    The plots span the full untrimmed motif, with the trimmed motif shaded.
</p>
<p>
    The hit-seqlet similarity is the cosine similarity between the additional-restricted-hits CWM and the seqlet CWM. 
    This statistic measures the similarity between hits that were missed by TF-MoDISco and the seqlets used to construct the motif.
</p>

<table class="wide_table">
    <thead>
        <tr>
            <th>Motif Name</th>
            {% if compute_recall %}
            <th class="num_col">Seqlet Recall</th>
            {% endif %}
            <th class="num_col">Hit-Seqlet CWM Similarity</th>
            <th class="num_col">Hits</th>
            {% if compute_recall %}
            <th class="num_col">Restricted Hits</th>
            {% endif %}
            {% if use_seqlets %}
            <th class="num_col">Seqlets</th>
            {% endif %}
            {% if compute_recall %}
            <th class="num_col">Hit/Seqlet Overlaps</th>
            <th class="num_col">Missed Seqlets</th>
            <th class="num_col">Additional Restricted Hits</th>
            {% endif %}
            <th class="cwm">Hit CWM (FC)</th>
            <th class="cwm">Hit CWM (RC)</th>
            <th class="cwm">TF-MoDISco CWM (FC)</th>
            <th class="cwm">TF-MoDISco CWM (RC)</th>
            {% if compute_recall %}
            <th class="cwm">Missed-Seqlet-Only CWM</th>
            <th class="cwm">Additional-Restricted-Hit CWM</th>
            {% endif %}
        </tr>
    </thead>
    <tbody>
        {% for item in report_data %}
        <tr>
            <td><code>{{ item.motif_name }}</code></td>
            {% if compute_recall %}
            <td class="num_col">{{ '%0.3f'| format(item.seqlet_recall|float) }}</td>
            {% endif %}
            <td class="num_col">{{ '%0.3f'| format(item.cwm_similarity|float) }}</td>
            <td class="num_col">{{ item.num_hits_total }}</td>
            {% if compute_recall %}
            <td class="num_col">{{ item.num_hits_restricted }}</td>
            {% endif %}
            {% if use_seqlets %}
            <td class="num_col">{{ item.num_seqlets }}</td>
            {% endif %}
            {% if compute_recall %}
            <td class="num_col">{{ item.num_overlaps }}</td>
            <td class="num_col">{{ item.num_seqlets_only }}</td>
            <td class="num_col">{{ item.num_hits_restricted_only }}</td>
            {% endif %}
            <td class="cwm"><img src="CWMs/{{ item.motif_name }}/hits_fc.svg" width="360"></td>
            <td class="cwm"><img src="CWMs/{{ item.motif_name }}/hits_rc.svg" width="360"></td>
            <td class="cwm"><img src="CWMs/{{ item.motif_name }}/modisco_fc.svg" width="360"></td>
            <td class="cwm"><img src="CWMs/{{ item.motif_name }}/modisco_rc.svg" width="360"></td>
            {% if compute_recall %}
            <td class="cwm"><img src="CWMs/{{ item.motif_name }}/seqlets_only.svg" width="360"></td>
            <td class="cwm"><img src="CWMs/{{ item.motif_name }}/hits_restricted_only.svg" width="360"></td>
            {% endif %}
        </tr>
        {% endfor %}
    </tbody>
</table>

{% if compute_recall %}

<h3>Seqlet-hit confusion matrix</h3>
<p>
    This heatmap shows the prevalence of motifs whose (untrimmed) hits overlap with TF-MoDISco seqlets of other motifs.
    The vertical axis shows the motif of the seqlet, while the horizontal axis shows the motif of the hit.
    The color intensity here represents an estimator of the expected number of bases of hit overlap per base of seqlet. 
</p>
<img src="seqlet_confusion.svg" width="780">

{% endif %}

<h2>Hit statistic distributions</h2>
<p>
    The following figures visualize the distribution of hit statistics across motifs and regions. 
</p>

<h3>Overall distribution of hit counts per region</h3>
<p>
    This plot shows the distribution of hit counts per region for any motif.
    The number of regions with no hits should be near zero. 
</p>
<img src="total_hit_distribution.svg" width="780">

<h3>Per-motif distributions of hit statistics</h3>
<p>
    These plots show the distribution of hit statistics for each motif, specifically:
    <ul>
        <li><strong>Hits Per Region</strong>: The number of hits per region for each motif</li>
        <li><strong>Hit Coefficient</strong>: The the distribution of hit coefficients for each motif</li>
        <li><strong>Hit Similarity</strong>: The distribution of hit-CWM similarity for each motif</li>
        <li><strong>Hit Importance</strong>: The distribution of the total importance of hits for each motif</li>
    </ul>
</p>
<table>
    <thead>
        <tr>
            <th>Motif Name</th>
            <th class="distplot">Hits Per Region</th>
            <th class="distplot">Hit Coefficient</th>
            <th class="distplot">Hit Similarity</th>
            <th class="distplot">Hit Importance</th>
        </tr>
    </thead>
    <tbody>
        {% for m in motif_names %}
        <tr>
            <td><code>{{ m }}</code></td>
            <td class="distplot"><img src="motif_hit_distributions/{{ m }}.svg" width="360"></td>
            <td class="distplot"><img src="motif_stat_distributions/{{ m }}_coefficients.svg" width="360"></td>
            <td class="distplot"><img src="motif_stat_distributions/{{ m }}_similarities.svg" width="360"></td>
            <td class="distplot"><img src="motif_stat_distributions/{{ m }}_importances.svg" width="360"></td>
        </tr>
        {% endfor %}
    </tbody>
</table>

<h3>Motif co-occurrence</h3>
<p>
    This heatmap shows the co-occurrence of motifs across regions.
    The color intensity here represents the cosine similarity between the motifs' occurrence across regions,
    where occurence is defined as the presence of a hit for a motif in a region.
</p>
<img src="motif_cooocurrence.svg" width="780">

</body>
</html>
